PurchasesTests/Purchasing/StoreKitRequestFetcherTests.swift

```diff
@@ -102,7 +101,7 @@ class StoreKitRequestFetcher: XCTestCase {
     
     func testOnlyCreatesOneRefreshRequest() {
         setupFetcher(fails: false)
-        expect(self.factory?.requests).to(haveCount(1))
+        expect(self.factory?.requests).toEventually(haveCount(1))
```

Comment:
here's the change in behavior I mentioned earlier. 
Note that the method is already async, so functionally it's the same. But the request won't be created synchronously.

Reply:
I'm wondering if we should add some extra tests to flex the `async` change. I'm paranoid that we have potential for concurrency bugs to pop up now that we're not using `@synchronized` to ensure atomicity like in `func finishReceiptRequest(_ request: SKRequest?) {` 
where we're setting:
```swift
self.receiptRefreshRequest = nil
...
self.receiptRefreshCompletionHandlers = []
```
within a dispatch async.
I guess I worry that `handler()` can be a closure that ends up calling this method asynchronously giving us a race condition.

Reply:
yeah, good point. I refactored and then missed the fact that the `handler()` might mess things up. 
In practice, the handler shouldn't be calling `finishReceiptRequest`, since it's only called from the `SKRequestDelegate` methods, which are called by the OS, and it's a private method so it won't be called from the outside. 
But this could and should be handled better. 
The advantage we had before was that we first locked, then modified the object, then called the handlers. 

We can't do exactly that with the queue, but we could make sure that the handlers are only called after we're done with the object, and done asynchronously. I'll do that, and then add tests

Reply:
Here's a helper, you can use this just as you would `@synchronized` and you could make it available across the project so if we need to `@synchronized` anything, we could do this!
"""
@inlinable
public func synchronized<R>(_ lock: Any, _ body: () throws -> R) rethrows -> R {
    objc_sync_enter(self)
    defer { objc_sync_exit(self) }
    return try body()
}
```
example:

```
synchronized(self) {
    print("doing this thing synchronized")
}
```
This solution is something modified from this post: https://jackmorris.xyz/2020/a-basic-reentrant-mutex/ because I didn't want to make a `Mutex` class and thought using `synchronized` was more elegant.

Reply:
I forgot to provide an update here, but I modified the method just after my comment to call the handlers _after_ the object is modified. Could you take a look? 

Thanks for posting the helper! 

The approach with `synchronized` is interesting! I assume the `self` in `objc_sync_enter(self)` should be `objc_sync_enter(lock)`? 

As someone used to obj-c I like the idea of retaining the syntax, but I'm a bit concerned about using a [concurrency mechanism designed for a different language](https://straypixels.net/swift-dictionary-locking/) - is it safe to use for copy-on-write objects? 

I'd honestly prefer sticking to queues for the sake of simplicity, since mutexes end up tripping me up often. Then again, maybe I haven't spent enough time with queues to get tripped up yet. 

What do you think?

Reply:
re: `lock`, yeah you're right, `self` should be `lock` 
Thanks for the link to that article, copy-on-write objects safety... I dunno! 
Going with queues in this case is probably safer. I worry about requiring a recursive lock here, but if you're sure this will never be called by the block supplied, then 👍